pkg/storage/snapshot/snapshot.go

```diff
@@ -675,43 +677,34 @@ func (ctrl *VMSnapshotController) updateSnapshotStatus(vmSnapshot *snapshotv1.Vi
 		}
 	}
 
+	canremoveFinalizer := true
 	if source != nil {
 		uid := source.UID()
 		vmSnapshotCpy.Status.SourceUID = &uid
+		canremoveFinalizer = !source.Locked()
 	}
 
-	content, err := ctrl.getContent(vmSnapshot)
-	if err != nil {
-		return err
-	}
-
-	if vmSnapshotDeleting(vmSnapshotCpy) {
-		// Enable the vmsnapshot to be deleted only in case it completed
-		// or after waiting until the content is deleted if needed
-		if !vmSnapshotProgressing(vmSnapshot) || contentDeletedIfNeeded(vmSnapshotCpy, content) {
-			controller.RemoveFinalizer(vmSnapshotCpy, vmSnapshotFinalizer)
-		}
-	} else {
-		// since no status subresource can update metadata and status
-		controller.AddFinalizer(vmSnapshotCpy, vmSnapshotFinalizer)
-
-		if content != nil && content.Status != nil {
-			// content exists and is initialized
-			vmSnapshotCpy.Status.VirtualMachineSnapshotContentName = &content.Name
-			vmSnapshotCpy.Status.CreationTime = content.Status.CreationTime
-			vmSnapshotCpy.Status.ReadyToUse = content.Status.ReadyToUse
-			vmSnapshotCpy.Status.Error = content.Status.Error
-		}
+	if content != nil && content.Status != nil {
+		// content exists and is initialized
+		vmSnapshotCpy.Status.VirtualMachineSnapshotContentName = &content.Name
+		vmSnapshotCpy.Status.CreationTime = content.Status.CreationTime
+		vmSnapshotCpy.Status.ReadyToUse = content.Status.ReadyToUse
+		vmSnapshotCpy.Status.Error = content.Status.Error
 	}
 
+	// terminal phase 1 - failed
 	if vmSnapshotDeadlineExceeded(vmSnapshotCpy) {
 		vmSnapshotCpy.Status.Phase = snapshotv1.Failed
 		updateSnapshotCondition(vmSnapshotCpy, newProgressingCondition(corev1.ConditionFalse, vmSnapshotDeadlineExceededError))
 		updateSnapshotCondition(vmSnapshotCpy, newFailureCondition(corev1.ConditionTrue, vmSnapshotDeadlineExceededError))
-	} else if vmSnapshotError(vmSnapshotCpy) != nil {
-		updateSnapshotCondition(vmSnapshotCpy, newProgressingCondition(corev1.ConditionFalse, "In error state"))
-		updateSnapshotCondition(vmSnapshotCpy, newReadyCondition(corev1.ConditionFalse, "Error"))
```

Comment:
why in case of error you don't want the ready condition to also represent that there was an error?